### PR TITLE
Adding Nek5000 to Spack

### PR DESCRIPTION
**Pull request by thilinarmtb**
This PR will  add Nek5000 - a fast high order spectral element CFD solver
to Spack package manager as part of CEED software suite.

This is still a work in progress, so please don't merge.

## Current status

This PR can successfully install Nek5000 tools and load tools + makenek file
with `module load`.

## TODO

- [x] Add the remaining tools as well.
- [x] Set MPI, PROFILING and VISIT parameters in `makenek` appropriately
- [x]  See what other CEED related constraints should be satisfied by this PR.
- [x] Test with all the versions

### CODE REVIEW

File: var/spack/repos/builtin/packages/nek5000/package.py [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+# WITHOUT ANY WARRANTY; without even the IMPLIED WARRANTY OF
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the terms and
+# conditions of the GNU Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public
+# License along with this program; if not, write to the Free Software
+# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
+##############################################################################
+#
+# This is a template package file for Spack.  We've put "FIXME"
```
Comment (davydden) [inline var/spack/repos/builtin/packages/nek5000/package.py]: please remove those left-over comments.

File: var/spack/repos/builtin/packages/nek5000/package.py [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+    """A fast and scalable high-order solver for computational fluid
+       dynamics"""
+
+    homepage = "https://nek5000.mcs.anl.gov/"
+    url      = "https://github.com/Nek5000/Nek5000"
+
+    tags = ['cfd', 'flow', 'hpc', 'solver', 'navier-stokes', \
+                                         'spectral-elements', 'fluid']
+
+    version('17.0.0-beta2', git = 'https://github.com/Nek5000/Nek5000.git', \
```
Comment (davydden) [inline var/spack/repos/builtin/packages/nek5000/package.py]: you don't need `\` and below.

File: var/spack/repos/builtin/packages/nek5000/package.py [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+    version('develop'     , git = 'https://github.com/Nek5000/Nek5000.git', \
+           branch='master')
+
+    variant('mpi'    , default=True, description='Build with MPI.'    )
+    variant('genbox' , default=True, description='Build genbox tool.' )
+    variant('genmap' , default=True, description='Build genmap tool.' )
+    variant('prenek' , default=True, description='Build prenek tool.' )
+    variant('postnek', default=True, description='Build psotnek tool.')
+
+    depends_on('mpi', when="+mpi", type=('build'))
```
Comment (davydden) [inline var/spack/repos/builtin/packages/nek5000/package.py]: most likely it's `build+link` (which is default, so you don't need to specify `type`), not only `build`.

File: var/spack/repos/builtin/packages/nek5000/package.py [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+            # Update the makenek to use correct compilers and
+            # Nek5000 source. Nek5000 source is manually copied.
+            filter_file(r'^F77\s*=.*', 'F77=\"' + F77 + '\"',  'makenek')
+            filter_file(r'^CC\s*=.*' , 'CC=\"'  + CC  + '\"',  'makenek')
+            filter_file(r'SOURCE_ROOT\s*=\"\$H.*',  'SOURCE_ROOT=\"'  + \
+                                             prefix.bin.Nek5000 + '\"',  'makenek')
+            install('makenek', installDir)
+
+            #FIXME "Not portable across platforms"
+            call('cp -r ../../Nek5000 '  + installDir, shell=True)
```
Comment (junghans) [inline var/spack/repos/builtin/packages/nek5000/package.py]: `install_tree()`?

File: var/spack/repos/builtin/packages/nek5000/package.py [hunk 5]
```diff
@@ -0,0 +1,123 @@
+##############################################################################
+# Copyright (c) 2013-2017, Lawrence Livermore National Security, LLC.
+# Produced at the Lawrence Livermore National Laboratory.
+#
+# This file is part of Spack.
+# Created by Todd Gamblin, tgamblin@llnl.gov, All rights reserved.
+# LLNL-CODE-647188
+#
+# For details, see https://github.com/llnl/spack
```
Comment (junghans) [inline var/spack/repos/builtin/packages/nek5000/package.py]: Update url to `spack/spack`!

**Comment by thilinarmtb** [comment]
Thank you very much for the feedback @davydden and @junghans !! I have updated the PR.

I am not ready to merge yet. One the other TODO item is done, we can merge this.

**Comment by thilinarmtb** [comment]
Also, looks like formatting checks fail. Is there a script you use
to enforce formatting?

Thanks in advance.

**Comment by davydden** [comment]
> Is there a script you use
to enforce formatting?

i don't think there is one, but you can run the tests locally `$ spack flake8`.

**Comment by thilinarmtb** [comment]
Thanks Denis and Christoph for the replies. I fixed the flake errors.

**Comment by junghans** [comment]
Is this still WIP?

**Comment by tgamblin** [comment]
Still has todos per comment above.  Should other CEED folks like @tzanio or @BarrySmith review this?

**Comment by thilinarmtb** [comment]
I added support for MPI, PROFILING and VISIT variants for Nek5000.

If we install Nek5000 with Visit enabled, I need to find the visit installation
directory in spack and update a variable in make script for Nek5000. 

How can I find the installation path of a spack package?

File: var/spack/repos/builtin/packages/nek5000/package.py [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+    depends_on('visit', when="+vist")
+
+    @run_before('install')
+    def fortran_check(self):
+        if not self.compiler.fc:
+            msg = 'Cannot build Nek5000 without a Fortran compiler.'
+            raise RuntimeError(msg)
+
+    @run_before('install')
+    def c_check(self):
```
Comment (thilinarmtb) [inline var/spack/repos/builtin/packages/nek5000/package.py]: Is this check needed? 
Does `spack` guarantee that there will always be a C compiler?

Comment (tgamblin) [reply var/spack/repos/builtin/packages/nek5000/package.py]: You can set up a `compilers.yaml` without one, but I know of no configurations that would result in a Fortran compiler being configured without a C compiler.  I think yo8u can safely take this out.

Comment (thilinarmtb) [reply var/spack/repos/builtin/packages/nek5000/package.py]: I will remove the C check.

**Comment by thilinarmtb** [comment]
Thanks @tgamblin for the replies. I removed the C check.

Nek5000 can depend on Visit and when it does I need to find the visit
installation directory in spack and update a variable in make script for
Nek5000. 

How can I find the installation path of a `spack` package?

After I fix this, this is good to be merged.

**Comment by thilinarmtb** [comment]
Any update on how to find `Visit` installation path?

**Comment by thilinarmtb** [comment]
I think I figured out how to access visit binary directory. This is what I have now:
[visit_bin_dir](https://github.com/spack/spack/pull/6153/files#diff-7d052b4bc7b2c09eb2666e546fd6d2fcR116)

Is that okay?

Also, I can't seem to install visit with `spack install visit`. I created an issue for this:
https://github.com/spack/spack/issues/6406

**Comment by thilinarmtb** [comment]
Is there a way to read a single parameter during `spack install`?

I need to read one parameter, `MAXNEL` which is a compile time
parameter to build some of the tools associated with `nek5000`.

Using a plain `input` in python does not work as the `stdout` is
redirected to a log file. How can I do this?

**Comment by thilinarmtb** [comment]
Just a kind reminder about my question above. @davydden, @tgamblin @junghans 

Thanks in advance !!

**Comment by junghans** [comment]
It helps to remove the `[WIP]` tag! ;-)

**Comment by junghans** [comment]
And yes, see the `maxvars` variant of `netcdf`.

**Comment by thilinarmtb** [comment]
Thank you very much @junghans for the reply !!

I made the necessary changes. I am okay to merge as it is if this looks fine.
I can make additional required changes as separate PRs if necessary.

**Comment by thilinarmtb** [comment]
I can cleanup the git history a little bit before getting merge. Let me know
if that is necessary.

File: var/spack/repos/builtin/packages/nek5000/package.py [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+    # MPI, Profiling and Visit variants
+    variant('mpi',       default=True, description='Build with MPI.')
+    variant('profiling', default=True, description='Build with profiling data.')
+    variant('visit',     default=False, description='Build with Visit.')
+
+    # Variant for MAXNEL, we need to read this from user
+    variant(
+        'MAXNEL',
+        default=150000,
+        description='Defines the maximum variables of NetCDF files.',
```
Comment (junghans) [inline var/spack/repos/builtin/packages/nek5000/package.py]: NetCDF?

Comment (thilinarmtb) [reply var/spack/repos/builtin/packages/nek5000/package.py]: Forgot to edit it. Fixed. Thanks !!

**Pull request merged**